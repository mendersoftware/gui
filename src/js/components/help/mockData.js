// Copyright 2020 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
export const helpProps = {
  docsVersion: undefined,
  isHosted: false,
  menderVersion: 'master',
  isEnterprise: true,
  tenantCapabilities: {
    hasDeviceConfig: true,
    hasMonitor: true
  },
  tokens: [],
  versions: {
    repos: {
      mender: '3.3.0',
      'mender-artifact': '3.8.0',
      'mender-binary-delta': '1.3.1',
      'mender-cli': '1.8.0',
      'mender-configure-module': '1.0.4',
      'mender-connect': '2.0.2',
      'mender-convert': '3.0.0',
      'mender-gateway': '1.0.0',
      'monitor-client': '1.2.0'
    },
    releaseDate: '2003-01-01'
  }
};
