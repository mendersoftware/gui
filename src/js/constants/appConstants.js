module.exports = {
  SET_SNACKBAR: 'SET_SNACKBAR',
  RECEIVED_HOSTED_LINKS: 'RECEIVED_HOSTED_LINKS',
  SET_LOCAL_IPADDRESS: 'SET_LOCAL_IPADDRESS',
  PLANS: {
    os: 'Mender Open Source',
    professional: 'Mender Professional',
    enterprise: 'Mender Enterprise'
  }
};
