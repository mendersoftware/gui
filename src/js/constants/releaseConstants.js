// Copyright 2019 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
export const ARTIFACT_GENERATION_TYPE = { SINGLE_FILE: 'single_file' };
export const ARTIFACTS_REMOVED_ARTIFACT = 'ARTIFACTS_REMOVED_ARTIFACT';
export const ARTIFACTS_SET_ARTIFACT_URL = 'ARTIFACTS_SET_ARTIFACT_URL';
export const UPDATED_ARTIFACT = 'UPDATED_ARTIFACT';
export const RECEIVE_ARTIFACTS = 'RECEIVE_ARTIFACTS';
export const RECEIVE_RELEASE = 'RECEIVE_RELEASE';
export const RECEIVE_RELEASES = 'RECEIVE_RELEASES';
export const RELEASE_REMOVED = 'RELEASE_REMOVED';
export const SHOW_REMOVE_DIALOG = 'SHOW_REMOVE_DIALOG';
export const SELECTED_ARTIFACT = 'SELECTED_ARTIFACT';
export const SELECTED_RELEASE = 'SELECTED_RELEASE';
export const SET_RELEASES_LIST_STATE = 'SET_RELEASES_LIST_STATE';
export const rootfsImageVersion = 'rootfs-image.version';

export const softwareTitleMap = { [rootfsImageVersion]: { title: 'Root filesystem', priority: 0, key: rootfsImageVersion } };

export const ALL_RELEASES = 'All releases';
