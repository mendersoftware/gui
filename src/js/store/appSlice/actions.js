// Copyright 2019 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
import { createAsyncThunk } from '@reduxjs/toolkit';
import { getOfflineThresholdSettings } from '@store/selectors';
import Cookies from 'universal-cookie';

import { actions, sliceName } from '.';
import { deepCompare, extractErrorMessage, getComparisonCompatibleVersion } from '../../helpers';
import GeneralApi from '../api/general-api';
import { searchDevices } from '../thunks';
import { getFeatures, getSearchState } from './selectors';

const cookies = new Cookies();

/*
  General
*/
export const setFirstLoginAfterSignup = createAsyncThunk(`${sliceName}/setFirstLoginAfterSignup`, (firstLoginAfterSignup, { dispatch }) => {
  cookies.set('firstLoginAfterSignup', !!firstLoginAfterSignup, { maxAge: 60, path: '/', domain: '.mender.io', sameSite: false });
  dispatch(actions.setFirstLoginAfterSignup(!!firstLoginAfterSignup));
});

const dateFunctionMap = {
  getDays: 'getDate',
  setDays: 'setDate'
};
export const setOfflineThreshold = createAsyncThunk(`${sliceName}/setOfflineThreshold`, (_, { dispatch, getState }) => {
  const { interval, intervalUnit } = getOfflineThresholdSettings(getState());
  const today = new Date();
  const intervalName = `${intervalUnit.charAt(0).toUpperCase()}${intervalUnit.substring(1)}`;
  const setter = dateFunctionMap[`set${intervalName}`] ?? `set${intervalName}`;
  const getter = dateFunctionMap[`get${intervalName}`] ?? `get${intervalName}`;
  today[setter](today[getter]() - interval);
  let value;
  try {
    value = today.toISOString();
  } catch {
    return Promise.resolve(dispatch(actions.setSnackbar('There was an error saving the offline threshold, please check your settings.')));
  }
  return Promise.resolve(dispatch(actions.setOfflineThreshold(value)));
});

const versionRegex = new RegExp(/\d+\.\d+/);
const getLatestRelease = thing => {
  const latestKey = Object.keys(thing)
    .filter(key => versionRegex.test(key))
    .sort()
    .reverse()[0];
  return thing[latestKey];
};

const repoKeyMap = {
  integration: 'Integration',
  mender: 'Mender-Client',
  'mender-artifact': 'Mender-Artifact'
};

const deductSaasState = (latestRelease, guiTags, saasReleases) => {
  const latestGuiTag = guiTags.length ? guiTags[0].name : '';
  const latestSaasRelease = latestGuiTag.startsWith('saas-v') ? { date: latestGuiTag.split('-v')[1].replaceAll('.', '-'), tag: latestGuiTag } : saasReleases[0];
  return latestSaasRelease.date > latestRelease.release_date ? latestSaasRelease.tag : latestRelease.release;
};

export const getLatestReleaseInfo = createAsyncThunk(`${sliceName}/getLatestReleaseInfo`, (_, { dispatch, getState }) => {
  if (!getFeatures(getState()).isHosted) {
    return Promise.resolve();
  }
  return Promise.all([GeneralApi.get('/versions.json'), GeneralApi.get('/tags.json')])
    .catch(err => {
      console.log('init error:', extractErrorMessage(err));
      return Promise.resolve([{ data: {} }, { data: [] }]);
    })
    .then(([{ data }, { data: guiTags }]) => {
      if (!guiTags.length) {
        return Promise.resolve();
      }
      const { releases, saas } = data;
      const latestRelease = getLatestRelease(getLatestRelease(releases));
      const { latestRepos, latestVersions } = latestRelease.repos.reduce(
        (accu, item) => {
          if (repoKeyMap[item.name]) {
            accu.latestVersions[repoKeyMap[item.name]] = getComparisonCompatibleVersion(item.version);
          }
          accu.latestRepos[item.name] = getComparisonCompatibleVersion(item.version);
          return accu;
        },
        { latestVersions: { ...getState().app.versionInformation }, latestRepos: {} }
      );
      const info = deductSaasState(latestRelease, guiTags, saas);
      return Promise.resolve(
        dispatch(
          actions.setVersionInformation({
            ...latestVersions,
            backend: info,
            GUI: info,
            latestRelease: {
              releaseDate: latestRelease.release_date,
              repos: latestRepos
            }
          })
        )
      );
    });
});

export const setSearchState = createAsyncThunk(`${sliceName}/setSearchState`, (searchState, { dispatch, getState }) => {
  const currentState = getSearchState(getState());
  let nextState = {
    ...currentState,
    ...searchState,
    sort: {
      ...currentState.sort,
      ...searchState.sort
    }
  };
  let tasks = [];
  // eslint-disable-next-line no-unused-vars
  const { isSearching: currentSearching, deviceIds: currentDevices, searchTotal: currentTotal, ...currentRequestState } = currentState;
  // eslint-disable-next-line no-unused-vars
  const { isSearching: nextSearching, deviceIds: nextDevices, searchTotal: nextTotal, ...nextRequestState } = nextState;
  if (nextRequestState.searchTerm && !deepCompare(currentRequestState, nextRequestState)) {
    nextState.isSearching = true;
    tasks.push(
      dispatch(searchDevices(nextState))
        .then(results => {
          const searchResult = results[results.length - 1];
          return dispatch(actions.setSearchState({ ...searchResult, isSearching: false }));
        })
        .catch(() => dispatch(actions.setSearchState({ isSearching: false, searchTotal: 0 })))
    );
  }
  tasks.push(dispatch(actions.setSearchState(nextState)));
  return Promise.all(tasks);
});
