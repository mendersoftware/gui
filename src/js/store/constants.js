export * from './appSlice/constants';
export * from './commonConstants';
export * from './deploymentsSlice/constants';
export * from './devicesSlice/constants';
export * from './monitorSlice/constants';
export * from './onboardingSlice/constants';
export * from './organizationSlice/constants';
export * from './releasesSlice/constants';
export * from './usersSlice/constants';
