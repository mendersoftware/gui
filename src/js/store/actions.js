import { actions as appActions } from './appSlice';
import { actions as deploymentActions } from './deploymentsSlice';
import { actions as devicesActions } from './devicesSlice';
import { actions as monitorActions } from './monitorSlice';
import { actions as onboardingActions } from './onboardingSlice';
import { actions as organizationActions } from './organizationSlice';
import { actions as releaseActions } from './releasesSlice';
import { actions as userActions } from './usersSlice';

export default {
  ...appActions,
  ...deploymentActions,
  ...devicesActions,
  ...monitorActions,
  ...onboardingActions,
  ...organizationActions,
  ...releaseActions,
  ...userActions
};
