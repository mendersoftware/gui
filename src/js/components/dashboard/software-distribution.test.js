// Copyright 2019 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
import React from 'react';

import { act, waitFor } from '@testing-library/react';

import { defaultState, undefineds } from '../../../../tests/mockData';
import { render } from '../../../../tests/setupTests';
import * as DeviceActions from '../../actions/deviceActions';
import { TIMEOUTS, chartTypes } from '../../constants/appConstants';
import { rootfsImageVersion } from '../../constants/releaseConstants';
import SoftwareDistribution from './software-distribution';

const preloadedState = {
  ...defaultState,
  devices: {
    ...defaultState.devices,
    reports: [
      {
        items: [
          { key: 'something', count: 10 },
          { key: 'somethingMore', count: 20 }
        ],
        otherCount: 12,
        total: 42
      },
      {
        items: [
          { key: 'something', count: 10 },
          { key: 'somethingMore', count: 20 }
        ],
        otherCount: 12,
        total: 42
      }
    ]
  },
  users: {
    ...defaultState.users,
    globalSettings: {
      ...defaultState.users.globalSettings,
      [defaultState.users.currentUser]: {
        ...defaultState.users.globalSettings[defaultState.users.currentUser],
        reports: [
          { group: Object.keys(defaultState.devices.groups.byId)[0], attribute: 'artifact_name', type: 'distribution', chartType: chartTypes.pie.key },
          { group: Object.keys(defaultState.devices.groups.byId)[1], software: rootfsImageVersion, type: 'distribution', chartType: chartTypes.bar.key }
        ]
      }
    }
  }
};

const reportsSpy = jest.spyOn(DeviceActions, 'getReportsDataWithoutBackendSupport');

describe('Devices Component', () => {
  it('renders correctly', async () => {
    const ui = <SoftwareDistribution />;

    const { baseElement, rerender } = render(ui, { preloadedState });
    await act(async () => {
      jest.runAllTimers();
      jest.runAllTicks();
      return new Promise(resolve => resolve(), TIMEOUTS.threeSeconds);
    });
    await waitFor(() => expect(reportsSpy).toHaveBeenCalled());
    await waitFor(() => rerender(ui));
    const view = baseElement.firstChild;
    expect(view).toMatchSnapshot();
    expect(view).toEqual(expect.not.stringMatching(undefineds));
    reportsSpy.mockClear();
  });

  it('renders correctly for enterprise', async () => {
    const testState = {
      ...preloadedState,
      app: {
        ...preloadedState.app,
        features: {
          ...preloadedState.app.features,
          isEnterprise: true
        }
      }
    };
    const ui = <SoftwareDistribution />;
    const { baseElement, rerender } = render(ui, { preloadedState: testState });
    await act(async () => {
      jest.runAllTimers();
      jest.runAllTicks();
      return new Promise(resolve => resolve(), TIMEOUTS.threeSeconds);
    });
    await waitFor(() => expect(reportsSpy).toHaveBeenCalled());
    await waitFor(() => rerender(ui));
    const view = baseElement.firstChild;
    expect(view).toMatchSnapshot();
    expect(view).toEqual(expect.not.stringMatching(undefineds));
  });
});
