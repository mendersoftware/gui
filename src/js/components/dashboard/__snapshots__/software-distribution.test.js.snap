// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Devices Component renders correctly 1`] = `
.emotion-0 {
  gap: 16px;
}

.emotion-1 {
  max-width: 100%;
  font-family: Lato,sans-serif;
  font-size: 0.7545rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  height: 32px;
  color: rgba(10, 10, 11, 0.78);
  background-color: rgba(0, 0, 0, 0.08);
  border-radius: 16px;
  white-space: nowrap;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  cursor: unset;
  outline: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: 0;
  padding: 0;
  vertical-align: middle;
  box-sizing: border-box;
  background-color: #f7fafb;
  color: rgba(0, 0, 0, 0.38);
  text-transform: uppercase;
  background-color: #f7fafb;
  color: rgba(0, 0, 0, 0.38);
  text-transform: uppercase;
}

.emotion-1.Mui-disabled {
  opacity: 0.38;
  pointer-events: none;
}

.emotion-1 .MuiChip-avatar {
  margin-left: 5px;
  margin-right: -6px;
  width: 24px;
  height: 24px;
  color: #bcbcbc;
  font-size: 0.6964rem;
}

.emotion-1 .MuiChip-avatarColorPrimary {
  color: #fff;
  background-color: rgb(35, 85, 94);
}

.emotion-1 .MuiChip-avatarColorSecondary {
  color: #fff;
  background-color: rgb(65, 10, 46);
}

.emotion-1 .MuiChip-avatarSmall {
  margin-left: 4px;
  margin-right: -4px;
  width: 18px;
  height: 18px;
  font-size: 0.5804rem;
}

.emotion-1 .MuiChip-icon {
  margin-left: 5px;
  margin-right: -6px;
  color: #bcbcbc;
}

.emotion-1 .MuiChip-deleteIcon {
  -webkit-tap-highlight-color: transparent;
  color: rgba(10, 10, 11, 0.26);
  font-size: 22px;
  cursor: pointer;
  margin: 0 5px 0 -6px;
}

.emotion-1 .MuiChip-deleteIcon:hover {
  color: rgba(10, 10, 11, 0.4);
}

.emotion-1:hover {
  font-weight: bold;
}

.emotion-1:hover {
  font-weight: bold;
}

.emotion-2 {
  overflow: hidden;
  text-overflow: ellipsis;
  padding-left: 12px;
  padding-right: 12px;
  white-space: nowrap;
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.3929rem;
  color: #337a87;
}

.emotion-3 iconButton {
  margin-right: 8px;
}

.emotion-3.read {
  color: rgba(0, 0, 0, 0.38);
}

.emotion-4 {
  position: absolute;
  top: -5px;
  bottom: 0;
  left: -5px;
  right: -5px;
  border: 1px dashed #337a87;
  border-radius: 50%;
}

.emotion-4.read {
  border-color: rgba(0, 0, 0, 0.38);
}

.emotion-5 {
  font-size: 1rem;
  cursor: pointer;
}

.emotion-6 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.3929rem;
}

.emotion-6 iconButton {
  margin-right: 8px;
}

<div>
  <div
    class="dashboard margin-bottom-large"
  >
    <div
      class="widget"
    >
      <div
        class="flexbox center-aligned  emotion-0"
        style="align-items: end;"
      >
        <div
          class=""
          data-mui-internal-clone-element="true"
        >
          <div
            class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault emotion-1"
          >
            <span
              class="MuiChip-label MuiChip-labelMedium emotion-2"
            >
              Enterprise
            </span>
          </div>
        </div>
        <div
          class=""
          data-mui-internal-clone-element="true"
        >
          <div
            class="relative"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium  emotion-3"
              data-testid="HelpIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 17h-2v-2h2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25"
              />
            </svg>
            <div
              class="emotion-4 "
            />
          </div>
        </div>
      </div>
      <div
        class="flexbox centered muted emotion-5"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-6"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
          />
        </svg>
        <span
          class="emotion-5"
        >
          add a widget
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Devices Component renders correctly for enterprise 1`] = `
.emotion-0 {
  min-height: 30px;
}

.emotion-0 .MuiSvgIcon-root {
  margin-left: 8px;
}

.emotion-1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.3929rem;
}

.emotion-1 iconButton {
  margin-right: 8px;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.3929rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  padding: 5px;
  font-size: 1.0446rem;
  font-size: 1.2rem;
  color: rgba(0, 0, 0, 0.54);
}

.emotion-2::-moz-focus-inner {
  border-style: none;
}

.emotion-2.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-2 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-2:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-2:hover {
    background-color: transparent;
  }
}

.emotion-2.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.1607rem;
}

.emotion-3 iconButton {
  margin-right: 8px;
}

.emotion-4 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

.emotion-8 {
  display: grid;
  grid-template-columns: 200px 1fr;
  -webkit-column-gap: 16px;
  column-gap: 16px;
  margin-bottom: 16px;
}

.emotion-8>.flexbox.column>* {
  height: 20px;
}

.emotion-8 .barchart .MuiLinearProgress-root {
  background-color: #f7f7f7;
}

.emotion-8 .barchart .color-f7f7f7 .MuiLinearProgress-barColorPrimary {
  background-color: #f7f7f7;
}

.emotion-8 .barchart .color-337a87 .MuiLinearProgress-barColorPrimary {
  background-color: #337a87;
}

.emotion-8 .barchart .color-a31773 .MuiLinearProgress-barColorPrimary {
  background-color: #a31773;
}

.emotion-8 .barchart .color-00859e .MuiLinearProgress-barColorPrimary {
  background-color: #00859e;
}

.emotion-8 .barchart .color-14cfda .MuiLinearProgress-barColorPrimary {
  background-color: #14cfda;
}

.emotion-8 .barchart .color-9bfff0 .MuiLinearProgress-barColorPrimary {
  background-color: #9bfff0;
}

.emotion-8 .barchart .color-d5d5d5 .MuiLinearProgress-barColorPrimary {
  background-color: #d5d5d5;
}

.emotion-9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: grid;
  grid-template-columns: 1fr max-content;
  -webkit-column-gap: 16px;
  column-gap: 16px;
}

.emotion-9.indicating {
  grid-template-columns: min-content 1fr max-content;
  -webkit-column-gap: 8px;
  column-gap: 8px;
}

.emotion-10 {
  position: relative;
  overflow: hidden;
  display: block;
  height: 4px;
  z-index: 0;
  background-color: rgb(177, 204, 209);
}

@media print {
  .emotion-10 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-11 {
  width: 100%;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  -webkit-transition: -webkit-transform .4s linear;
  transition: transform .4s linear;
  transform-origin: left;
  background-color: #337a87;
}

.emotion-12 {
  gap: 16px;
}

.emotion-13 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.3929rem;
  color: #337a87;
}

.emotion-13 iconButton {
  margin-right: 8px;
}

.emotion-13.read {
  color: rgba(0, 0, 0, 0.38);
}

.emotion-14 {
  position: absolute;
  top: -5px;
  bottom: 0;
  left: -5px;
  right: -5px;
  border: 1px dashed #337a87;
  border-radius: 50%;
}

.emotion-14.read {
  border-color: rgba(0, 0, 0, 0.38);
}

.emotion-15 {
  font-size: 1rem;
  cursor: pointer;
}

<div>
  <div
    class="dashboard margin-bottom-large"
  >
    <div
      class="widget chart-widget"
    >
      <div
        class="margin-bottom-small"
      >
        <div
          class="flexbox space-between margin-bottom-small"
        >
          <div
            class="flexbox center-aligned emotion-0"
          >
            Software distribution
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-1"
              data-testid="BarChartIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M4 9h4v11H4zm12 4h4v7h-4zm-6-9h4v16h-4z"
              />
            </svg>
          </div>
          <div
            class="flexbox center-aligned"
            style="z-index: 1;"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall emotion-2"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall emotion-3"
                data-testid="SettingsIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6"
                />
              </svg>
              <span
                class="MuiTouchRipple-root emotion-4"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall emotion-2"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall emotion-3"
                data-testid="ClearIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root emotion-4"
              />
            </button>
          </div>
        </div>
        <div
          class="flexbox space-between slightly-smaller"
        >
          <div>
            artifact_name
          </div>
          <div>
            All devices
          </div>
        </div>
      </div>
      <div
        class="emotion-8"
      >
        <div
          class="flexbox column"
        >
          <div
            class="clickable emotion-9 "
            title="raspotifyInstaller"
          >
            <div
              class="text-overflow"
            >
              raspotifyInstaller
            </div>
            <div>
              2
            </div>
          </div>
        </div>
        <div
          class="flexbox column"
        >
          <div
            class="clickable flexbox column barchart"
            style="justify-content: center;"
            title="raspotifyInstaller"
          >
            <span
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="0"
              class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate color-337a87 emotion-10"
              role="progressbar"
              style="height: 8px;"
            >
              <span
                class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate emotion-11"
                style="transform: translateX(-100%);"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="widget"
    >
      <div
        class="flexbox center-aligned  emotion-12"
        style="align-items: end;"
      >
        <div
          class=""
          data-mui-internal-clone-element="true"
        >
          <div
            class="relative"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium  emotion-13"
              data-testid="HelpIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 17h-2v-2h2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25"
              />
            </svg>
            <div
              class="emotion-14 "
            />
          </div>
        </div>
      </div>
      <div
        class="flexbox centered muted emotion-15"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-1"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
          />
        </svg>
        <span
          class="emotion-15"
        >
          add a widget
        </span>
      </div>
    </div>
  </div>
</div>
`;
